src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs

```diff
@@ -151,15 +155,34 @@ private Task ProcessStreamBindingFailure(HubConnectionContext connection, Stream
 
             var message = CompletionMessage.WithError(bindingFailureMessage.Id, errorString);
             Log.ClosingStreamWithBindingError(_logger, message);
-            connection.StreamTracker.Complete(message);
+            connection.StreamTracker.TryComplete(message);
+
+            // TODO: Send stream completion message to client when we add it
 
             return Task.CompletedTask;
         }
 
         private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessage message)
         {
-            Log.ReceivedStreamItem(_logger, message);
-            return connection.StreamTracker.ProcessItem(message);
+            var processTask = connection.StreamTracker.ProcessItem(message);
+            // Review: This is kind of ugly and relies on the fact that the exception will be thrown from the synchronous code path
```

Comment:
Why are we doing this?

Reply:
Currently if the client sends a stream item after the stream was closed we kill the connection. This is to prevent killing the connection.

Reply:
We should not be using async vs. sync exceptions for control flow. Even `Task<<Task>>` with inner vs outer exceptions for control is slightly better. Or maybe `bool StreamTracker.TryProcessItem(out Task)`.